pkg/controller/garbagecollector/garbage_collector.go

```diff
@@ -0,0 +1,421 @@
+// ensures that the garbage collector operates with a graph that is at least as
+// up to date as the notification is sent.
+type GarbageCollector struct {
+	restMapper meta.RESTMapper
+	// clientPool uses the regular dynamicCodec. We need it to update
+	// finalizers. It can be removed if we support patching finalizers.
+	clientPool dynamic.ClientPool
+	// garbage collector attempts to delete the items in attemptToDelete queue when the time is ripe.
+	attemptToDelete workqueue.RateLimitingInterface
+	// garbage collector attempts to orphan the dependents of the items in the attemptToOrphan queue, then deletes the items.
+	attemptToOrphan        workqueue.RateLimitingInterface
+	dependencyGraphBuilder *GraphBuilder
+	// used to register exactly once the rate limiter of the dynamic client
+	// used by the garbage collector controller.
+	registeredRateLimiter *RegisteredRateLimiter
+	// GC caches the owners that do not exist according to the API server.
+	absentOwnerCache *UIDCache
+}
+
+func NewGarbageCollector(metaOnlyClientPool dynamic.ClientPool, clientPool dynamic.ClientPool, mapper meta.RESTMapper, deletableResources map[schema.GroupVersionResource]struct{}) (*GarbageCollector, error) {
+	attemptToDelete := workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "garbage_collector_attempt_to_delete")
+	attemptToOrphan := workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "garbage_collector_attempt_to_orphan")
+	absentOwnerCache := NewUIDCache(500)
+	gc := &GarbageCollector{
+		clientPool:            clientPool,
+		restMapper:            mapper,
+		attemptToDelete:       attemptToDelete,
+		attemptToOrphan:       attemptToOrphan,
+		registeredRateLimiter: NewRegisteredRateLimiter(deletableResources),
+		absentOwnerCache:      absentOwnerCache,
+	}
+	gb := &GraphBuilder{
+		metaOnlyClientPool:                  metaOnlyClientPool,
+		registeredRateLimiterForControllers: NewRegisteredRateLimiter(deletableResources),
+		restMapper:                          mapper,
+		graphChanges:                        workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "garbage_collector_graph_changes"),
+		uidToNode: &concurrentUIDToNode{
+			RWMutex:   &sync.RWMutex{},
+			uidToNode: make(map[types.UID]*node),
+		},
+		attemptToDelete:  attemptToDelete,
+		attemptToOrphan:  attemptToOrphan,
+		absentOwnerCache: absentOwnerCache,
+	}
+	if err := gb.monitorsForResources(deletableResources); err != nil {
+		return nil, err
+	}
+	gc.dependencyGraphBuilder = gb
+
+	return gc, nil
+}
+
+func (gc *GarbageCollector) Run(workers int, stopCh <-chan struct{}) {
+	defer gc.attemptToDelete.ShutDown()
+	defer gc.attemptToOrphan.ShutDown()
+	defer gc.dependencyGraphBuilder.graphChanges.ShutDown()
+
+	glog.Infof("Garbage Collector: Initializing")
+	gc.dependencyGraphBuilder.startMonitors(stopCh)
+	if !cache.WaitForCacheSync(stopCh, gc.dependencyGraphBuilder.MonitorsSynced) {
```

Comment:
You are right, my memory served me wrong. It's used to prevent this flake: https://github.com/kubernetes/kubernetes/issues/28713#issuecomment-234826276.